### DISCUSSION TITLE
YOUR CleanArchitecture implementation

### DISCUSSION THREAD

**Discussion by ivanpaulovich** [ivanpaulovich/clean-architecture-manga#169] [Show and tell]
Hey all,

I am happy to see many developers inspired by the CleanArchitectureManga that started awesome projects following these principles and patterns. Cheers for you all 🍺!

For the last months I have been receiving e-mails with **design questions about YOUR projects**, I learned a lot with your challenges and suggestions! I do my best to answer them quickly.

:point_right: Now I'd like to ask you to introduce your project challenges to this topic, share what you are trying to accomplish then everyone that follows the CleanArchitectureManga will take the benefit of discussions. :point_left:

The best is that you will get a wide view of what people are doing out there. We are all crafters here!

**Comment by petrspelos** [comment]
Hello, there! 👋 
Before I even found out about this project (which influenced my design a great bit), I tried to implement a sensible architecture in a community Discord bot project I've started called Miunie.

Now with Discord bot, the interesting issue is that while the host has an entry point in one of our multiple UI clients, the actual bot commands have their own entry-point through a Discord wrapper. (think ASP.NET Controllers, but in a different assembly from the standard entry point).

<details>
  <summary>Click here for more details...</summary>
So we've instinctively went with a variant design to what @ivanpaulovich did with this repo.

![image](https://user-images.githubusercontent.com/21304428/75151481-13fda900-5707-11ea-8e6f-9c8188424beb.png)

Our core is basically Application and Domain layers in one. Instead of UseCases, we went with Services, but they're ultimately, the same thing.

Sorry for the horrible UML, but this is the relation between our Miunie.Discord and Miunie.Core. This along with the separation of front-ends allows us to have multiple GUI clients, and even replace the Discord layer, should we need a different wrapper (which we did, since the wrapper we were using got deleted by the author). Replacing it was quite easy.

In this UML, the entry point of the user command (not application) is the `ProfileCommand` and it then calls the `ProfileService` for a more or less classic Use case flow.

![image](https://user-images.githubusercontent.com/21304428/75151738-b1f17380-5707-11ea-8539-3d2a7546af81.png)
</details>

------

TL;DR: The project isn't perfect and we're trying to slowly move it to something that more resembles Manga, but at the same time, we're honestly at a pretty alright place for a 1 year old project.

- [Miunie project](https://github.com/control-net/Miunie/wiki/About-Miunie)

I hope I understood this correctly, if not, sorry. 😅

**Comment by leandrofagundes** [comment]
I'm implementing a project for a company and I can't show the source code, but believe me or not, every developer of the team knows you (I asked they to put a star for you haha)

**Comment by brunocasarotti** [comment]
First of all, great project, very well explained! Something that I am trying to accomplish is to consume messages from a queue like Rabbit. I believe a new layer acting like the webapi but consuming messages from a queue in our infrastructure and calling our use cases would work. Would be nice to have a sample of that and not be restrict to only synchronous calls.

**Comment by raschle** [comment]
I used this clean architecture implementation at work with great success. It's used in the following, private, projects:

**Backend for an automatic car park system with 99 car spaces**
- Communication with Siemens PLC (S7-1500)
- MSSQL database with entity framework core
- User terminals connected over Signalr
- RFID reader connected via TCP
- API to cash register / ticketing system

**Gateway for connecting industrial food mixer**
- Communication with Siemens PLC (S7-1500)
- API to Microsoft Navision for order handling and feedback of produced articles

Thanks again to Ivan for this great project and the coaching/help I received.

**Reply by Larry57** [reply]
This is very interresting. Are you using an OPC library for the PLC communication? Which is it?

**Reply by raschle** [reply]
Snap7, direct S7 communication, reading/writing to DBs. It's the most performant but a vendor lock on Siemens. OPC is the better/future-proof way to do it.
